### PR TITLE
Add validations for Personalize input and configurations

### PR DESCRIPTION
**Pull request by kulket**
### Description
This pull request adds following changes:
* Change scoring logic for Personalized ranking to consider ranks instead of scores as well as refactor overall scoring logic implementation
* Add validation for search request parameters as well as response processor configurations related to Personalization
* Add E2E tests in the form of unit tests for Personalize ranking
* Add user agent configuration to Personalize client to understand plugin adoption
 
### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Commits are signed as per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/search-processor/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Review by mishprs — APPROVED** [review]

File: src/main/java/org/opensearch/search/relevance/transformer/personalizeintelligentranking/reranker/impl/AmazonPersonalizedRankerImpl.java [hunk 1]
```diff
@@ -10,11 +10,13 @@
 import com.amazonaws.services.personalizeruntime.model.GetPersonalizedRankingRequest;
 import com.amazonaws.services.personalizeruntime.model.GetPersonalizedRankingResult;
 import com.amazonaws.services.personalizeruntime.model.PredictedItem;
+import org.apache.commons.lang3.StringUtils;
```
Comment (noCharger) [inline src/main/java/org/opensearch/search/relevance/transformer/personalizeintelligentranking/reranker/impl/AmazonPersonalizedRankerImpl.java]: Using `StringUtils` requires adding the Apache Commons Lang library as a dependency. While this is not a significant issue, it does introduce an additional external dependency that needs to be managed and potentially increases the project's size. Also, calling utility methods for simple string operations (only isEmpty is used) can add some overhead compared to inline operations.

File: src/main/java/org/opensearch/search/relevance/transformer/personalizeintelligentranking/utils/ValidationUtil.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+            throw ConfigurationUtils.newConfigurationException(processorType, processorTag, "campaign_arn", "invalid format for Personalize campaign arn");
+        }
+        // Validate IAM Role Arn for Personalize access
+        String iamRoleArn = config.getIamRoleArn();
+        if(!StringUtils.isEmpty(iamRoleArn) && !isValidCampaignOrRoleArn(iamRoleArn, "iam")) {
+            throw ConfigurationUtils.newConfigurationException(processorType, processorTag, "iam_role_arn", "invalid format for Personalize iam role arn");
+        }
+        // Validate Personalize recipe
+        if(!SUPPORTED_PERSONALIZE_RECIPES.contains(config.getRecipe())) {
+            throw ConfigurationUtils.newConfigurationException(processorType, processorTag, "recipe", "not supported recipe provided");
```
Comment (noCharger) [inline src/main/java/org/opensearch/search/relevance/transformer/personalizeintelligentranking/utils/ValidationUtil.java:48]: Please add test case on this exception.

File: src/main/java/org/opensearch/search/relevance/transformer/personalizeintelligentranking/utils/ValidationUtil.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+            Arn arnObj = Arn.fromString(arn);
+            String arnService = arnObj.getService();
+            if (!arnService.equals(expectedService)) {
+                return false;
+            }
+        } catch (IllegalArgumentException iae) {
+            return false;
+        }
+        return true;
+    }
```
Comment (noCharger) [inline src/main/java/org/opensearch/search/relevance/transformer/personalizeintelligentranking/utils/ValidationUtil.java]:
```suggestion
private static boolean isValidCampaignOrRoleArn(String arn, String expectedService) {
    try {
        Arn arnObj = Arn.fromString(arn);
        return arnObj.getService().equals(expectedService);
    } catch (IllegalArgumentException ignored) {
        return false;
    }
}
```

**Review by noCharger — COMMENTED** [review]
Overall looks good, have some small comments

File: src/main/java/org/opensearch/search/relevance/transformer/personalizeintelligentranking/utils/ValidationUtil.java (hunk 2 above)
Comment (kulket) [reply src/main/java/org/opensearch/search/relevance/transformer/personalizeintelligentranking/utils/ValidationUtil.java:48]: Already have a test case supporting this: Refer [here](https://github.com/opensearch-project/search-processor/pull/151/files#diff-c0a77f60b58a5d7677bc0e4593721ed3627f793aba45e6f5123ec55f707ffb3aR88).

File: src/main/java/org/opensearch/search/relevance/transformer/personalizeintelligentranking/utils/ValidationUtil.java (hunk 3 above)
Comment (kulket) [reply src/main/java/org/opensearch/search/relevance/transformer/personalizeintelligentranking/utils/ValidationUtil.java]: Thank you for suggestion. Will make this change.

File: src/main/java/org/opensearch/search/relevance/transformer/personalizeintelligentranking/reranker/impl/AmazonPersonalizedRankerImpl.java (hunk 1 above)
Comment (kulket) [reply src/main/java/org/opensearch/search/relevance/transformer/personalizeintelligentranking/reranker/impl/AmazonPersonalizedRankerImpl.java]: Sure. Will replace it with ``!=null ``and ``isEmpty()`` checks on the ``String`` object itself.

**Review by mingshl — APPROVED** [review]
LGTM

File: src/main/java/org/opensearch/search/relevance/transformer/personalizeintelligentranking/PersonalizeRankingResponseProcessor.java [hunk 4]
```diff
@@ -41,7 +42,7 @@ public class PersonalizeRankingResponseProcessor extends AbstractProcessor imple
 
     private static final Logger logger = LogManager.getLogger(PersonalizeRankingResponseProcessor.class);
 
-    public static final String TYPE = "personalize_ranking";
+    public static final String TYPE = "personalized_search_ranking";
```
Comment (noCharger) [inline src/main/java/org/opensearch/search/relevance/transformer/personalizeintelligentranking/PersonalizeRankingResponseProcessor.java:45]: Do we want to change the class name as well? src/main/java/org/opensearch/search/relevance/transformer/personalizeintelligentranking/PersonalizeRankingResponseProcessor.java

Also there are multiple names now: `personalizeintelligentranking`, `PersonalizeRanking`, `personalized_search_ranking`. Can we make it consistent?

File: src/main/java/org/opensearch/search/relevance/transformer/personalizeintelligentranking/utils/ValidationUtil.java (hunk 2 above)
Comment (noCharger) [reply src/main/java/org/opensearch/search/relevance/transformer/personalizeintelligentranking/utils/ValidationUtil.java:48]:
```
 Check warning on line 48 in src/main/java/org/opensearch/search/relevance/transformer/personalizeintelligentranking/utils/ValidationUtil.java

Codecov
/ codecov/patch
src/main/java/org/opensearch/search/relevance/transformer/personalizeintelligentranking/utils/ValidationUtil.java#L48
```

**Review by noCharger — COMMENTED** [review]
Overall looks good. Left some non-blocking comments. Can't wait to see it included in 2.9!

File: src/main/java/org/opensearch/search/relevance/transformer/personalizeintelligentranking/PersonalizeRankingResponseProcessor.java (hunk 4 above)
Comment (kulket) [reply src/main/java/org/opensearch/search/relevance/transformer/personalizeintelligentranking/PersonalizeRankingResponseProcessor.java:45]: Agree. We will address in next iterations to make the naming more consistent.

**Pull request merged**